### DISCUSSION TITLE
Potential bug: `Never` as invariant type-var

### DISCUSSION THREAD

**Discussion by Azureblade3808** [microsoft/pyright#4891] [General]
Look at this --
```python
from typing import Any, Generic, Never, TypeVar

T = TypeVar("T")

class C(Generic[T]): ...

def f(cond: Any, c_never: C[Never], c_object: C[object], c_int: C[int]):
    if cond:
        c_never = c_object
    elif cond:
        c_never = c_int
    elif cond:
        c_object = c_never  # Not reported.
    elif cond:
        c_object = c_int
    elif cond:
        c_int = c_never  # Not reported.
    elif cond:
        c_int = c_object
```
It is expected that every assignment in `f` gets reported as `T` is invariant and `Never` is a different type from `object` or `int`.  Mypy is giving a correct result: https://mypy-play.net/?mypy=latest&python=3.12&gist=f14c91ed1eb6b2bbf31153ccf8fd0d45 .

If the type-var is changed into co-variant or contra-variant, the result will be correct accordingly.

I haven't thought of a practical case affected by this yet, but maybe it's worth some check.

**Comment by erictraut** [comment]
Thanks, I agree this is a bug. It will be addressed in the next release.
